theano/sandbox/gpuarray/multinomial.py

```diff
@@ -0,0 +1,256 @@
+        self.odtype = odtype
+
+    def get_params(self, node):
+        return node.outputs[0].type.context
+
+    def c_headers(self):
+        return ['<numpy_compat.h>', 'gpuarray_helper.h']
+
+    def c_header_dirs(self):
+        return [os.path.dirname(__file__), pygpu.get_include()]
+
+    def make_node(self, pvals, unis):
+        assert pvals.dtype == 'float32'
+        assert unis.dtype == 'float32'
+        ctx_name = infer_context_name(pvals, unis)
+
+        pvals = as_gpuarray_variable(pvals, ctx_name)
+        unis = as_gpuarray_variable(unis, ctx_name)
+
+        if pvals.ndim != 2:
+            raise NotImplementedError('pvals ndim should be 2', pvals.ndim)
+        if unis.ndim != 1:
+            raise NotImplementedError('unis ndim should be 1', unis.ndim)
+        if self.odtype == 'auto':
+            odtype = pvals.dtype
+        else:
+            odtype = self.odtype
+        assert odtype == 'float32', odtype
+        if odtype != pvals.dtype:
+            raise NotImplementedError(
+                'GpuMultinomialFromUniform works only if '
+                'self.odtype == pvals.dtype', odtype, pvals.dtype)
+        br = (pvals.broadcastable[1], pvals.broadcastable[0])
+        out = GpuArrayType(broadcastable=br, dtype=odtype)()
+
+        return Apply(self, [pvals, unis], [out])
+
+    def gpu_kernels(self, node, name):
+        dtype = node.outputs[0].dtype
+        code = """
+KERNEL void k_multi_warp_multinomial(
+    const ga_size nb_multi,
+    const ga_size nb_outcomes,
+    GLOBAL_MEM float * global_pvals,
+    const ga_ssize pvals_row_stride,
+    const ga_ssize pvals_col_stride,
+    GLOBAL_MEM float * global_unis,
+    const ga_ssize unis_stride,
+    GLOBAL_MEM float * global_outs,
+    const ga_ssize outs_row_stride,
+    const ga_ssize outs_col_stride
+)
+{
+    // each thread takes care of one multinomial draw
+    int n = LDIM_0*GID_0 + LID_0;
+    if (n < nb_multi)
+    {
+        float cummul = 0.;
+        bool done = false;
+        const float unis_n = global_unis[n*unis_stride];
+        for (ga_size m = 0; m < nb_outcomes; ++m)
+        {
+            float current_out = 0.;
+            if (!done)
+            {
+                cummul += global_pvals[m * pvals_col_stride +
+                                       n * pvals_row_stride];
+                if (unis_n < cummul)
+                {
+                    current_out = 1.;
+                    done = true;
+                }
+            }
+            //write out transposed for speed.
+            global_outs[n * outs_col_stride +
+                        m * outs_row_stride] = current_out;
+        }
+    }
+}
+
+
+//KERNEL void k(GLOBAL_MEM %(ctype)s *a, ga_size n, ga_size m) {
+//    ga_size nb = n < m ? n : m;
+//    for (ga_size i = LID_0; i < nb; i += LDIM_0) {
+//        a[i*m + i] = %(write_a)s(1);
+//    }
+//}""" % dict(ctype=pygpu.gpuarray.dtype_to_ctype(dtype),
+              name=name, write_a=write_w(dtype))
```

Comment:
`write_a` and `ctype` are only used for the commented kernel.  `name` is not used anywhere.